[#2] Refactor permission policy approach

fenekku opened the pull request:
- Closes #2 
- Make run-tests.sh pass locally and on Travis
- Comment out Deposit and Files permissions until we discuss them
- Merging to dev but we should start merging to master soon.

fenekku commented:
I will investigate Travis tomorrow.

ppanero commented on .editorconfig:
```diff
@@ -20,7 +20,7 @@ charset = utf-8
 # isort plugin configuration
 known_first_party = invenio_records_permissions
-multi_line_output = 2
+multi_line_output = 5
```
Hmm... This comes from `cookiecutter-invenio-module`. If it should be changed, could you make a PR there? if its a local default I think we should avoid pushing them (maybe add to the .gitignore?)

ppanero commented on invenio_records_permissions/generators.py:
```diff
@@ -6,227 +6,192 @@
 
-class _AnyUser(_NeedClass):
+    def needs(self, **kwargs):
+        """Enabling low-level Needs."""
```
What do you mean by "low-level"?

ppanero replied:
What do you mean by "low-level"?

ppanero commented on invenio_records_permissions/generators.py:
```diff
@@ -6,227 +6,192 @@
 
-    def query_filter(self):
-        return Q('match_all')  # match all query
+    def query_filter(self, **kwargs):
```
We should give a brainstorm to this, since is not only `filters` what can be returned here. Maybe just `query`? or `query_params`, `query_operations`, `query_restrictions`...

ppanero replied:
We should give a brainstorm to this, since is not only `filters` what can be returned here. Maybe just `query`? or `query_params`, `query_operations`, `query_restrictions`...

ppanero commented on invenio_records_permissions/generators.py:
```diff
@@ -6,227 +6,192 @@
-
-class _Admin(_NeedClass):
+class Deny(Generator):
+    """Denies ALL users (except super users)."""
```
This is not fully correct, as per implementation. In addition Lars mentioned that Deny all should deny even superusers. Although now I am not sure if its the right way to go.

ppanero replied:
This is not fully correct, as per implementation. In addition Lars mentioned that Deny all should deny even superusers. Although now I am not sure if its the right way to go.

ppanero commented on invenio_records_permissions/generators.py:
```diff
@@ -6,227 +6,192 @@
-    def excludes(self, record):
-        pass
+        """Constructor."""
+        super(Admin, self).__init__()
```
Following the previous comment. `Deny` should deny all access, if we want to grant admin, we should only specify `Admin`. Keeping in mind `excludes` have priority over `needs`.

ppanero commented on invenio_records_permissions/generators.py:
```diff
@@ -6,227 +6,192 @@
 
 from elasticsearch_dsl.query import Q
 from flask import g
+from flask_login import current_user
```
Is this `import` needed?

ppanero replied:
Is this `import` needed?

ppanero commented on invenio_records_permissions/generators.py:
```diff
@@ -6,227 +6,192 @@
-        super(_RecordOwners, self).__init__()
+class RecordOwners(Generator):
+    """Allows record owners."""
 
```
Removing the constructor call to `super` would make this one not have the `exclude` initialize and would fail? Not sure about this behaviour. To be tested. However, for consistency and safety, shouldn't we have a constructor as for the others?

ppanero commented on invenio_records_permissions/generators.py:
```diff
@@ -6,227 +6,192 @@
 
-class _RecordOwners(_RecordNeedClass):
 
-    def __init__(self):
```
Removing the constructor call to super would make this one not have the exclude initialize and would fail? Not sure about this behaviour. To be tested. However, for consistency and safety, shouldn't we have a constructor as for the others?

fenekku replied:
We are good on that one: invenio-records-permissions/tests/test_generators.py:77 tests that `excludes` is defined for `RecordOwners`. The constructor and other methods (like `excludes`) are inherited from the parent.

ppanero commented on invenio_records_permissions/generators.py:
```diff
@@ -6,227 +6,192 @@
 
-def _is_restricted(record, *args, **kwargs):
-    if record:
-        # FIXME: this should be caster to boolean when loaded
```
Thanks for fixing this! Not sure what pass by my head by putting `json.loads`...

fenekku replied:
:+1:

ppanero commented on invenio_records_permissions/generators.py:
```diff
@@ -6,227 +6,192 @@
-                return Q('term', **{"deposits.owners": need.value})
-        return None
 
+class AnyUserIfPublic(Generator):
```
This looks good to me after the refactor. However, we should revisit when we deal with Files/Deposit because I believe that many parts are able to be generalized. Maybe add a #TODO

ppanero commented on invenio_records_permissions/permissions/base.py:
```diff
@@ -6,99 +6,108 @@
+# |-------------|------|----------|---------------|
+# |     read    |   x  |     x    |        x      |
+# |-------------|------|----------|---------------|
+# | read files  |   x  |     x    |               |
```
I beleive  in the long run search will be perforemed on files also so it should have a `query_filter` but we can just put a #TODO for now.

ppanero commented on .editorconfig:
```diff
@@ -20,7 +20,7 @@ charset = utf-8
 # isort plugin configuration
 known_first_party = invenio_records_permissions
-multi_line_output = 2
+multi_line_output = 5
```
Hmm... This comes from cookiecutter-invenio-module. If it should be changed, could you make a PR there? if its a local default I think we should avoid pushing them (maybe add to the .gitignore?)

fenekku replied:
:+1:  I will be making a PR for a couple of changes to that repository. 

The isort plugin configuration in `.editorconfig` is strangely enough being picked up by the isort command being run in tests! I wouldn't have expected that: thought `.editorconfig` was just for editors...

fenekku replied:
I have tried using `__all__` but it ended up being more clunky (at least I thought so). If this change in `multi_line_output` is really not desired, I can go back to `__all__`.

ppanero replied:
I understand the issues and I share the pain. However, personally I would not change that config here until the architects approve the change in the cookiecutter repo. Mainly to avoid inconsistencies across repos. Once its merge we can put a task and change it across repos.

ppanero commented on invenio_records_permissions/generators.py:
```diff
@@ -6,227 +6,192 @@
-    def excludes(self, record):
-        pass
+        """Constructor."""
+        super(Admin, self).__init__()
```
Following the previous comment. `Deny` should deny all access, if we want to grant admin, we should only specify `Admin`. Keeping in mind `excludes` have priority over `needs`.

fenekku replied:
see above

ppanero commented on invenio_records_permissions/generators.py:
```diff
@@ -6,227 +6,192 @@
-                return Q('term', **{"deposits.owners": need.value})
-        return None
 
+class AnyUserIfPublic(Generator):
```
This looks good to me after the refactor. However, we should revisit when we deal with Files/Deposit because I believe that many parts are able to be generalized. Maybe add a #TODO

fenekku replied:
agreed

ppanero commented on invenio_records_permissions/permissions/base.py:
```diff
@@ -6,99 +6,108 @@
+    """
 
+    # Deny all by default
+    can_list = [Deny()]
```
I understand why putting this config here. However, since the definition is an *action* singular, its not fully simple. I'm not fully convinced, but creating a config for this is over engineering since I don't find any other reason than "readability".

fenekku replied:
If  I understand correctly, you mean that we shouldn't have `Deny()` at all? I agree with that.

ppanero commented on invenio_records_permissions/permissions/base.py:
```diff
@@ -6,99 +6,108 @@
+    can_update = [Deny()]
+    can_delete = [Deny()]
+
+    def __init__(self, action, **over):
```
Shall we pass `*args` and ``**kwargs`` and then look for the `over` key in one of them? Just to give more flexibility in the future.

fenekku replied:
I wanted `**over` here for more clarity in our specific usage (`**kwargs` is too generic): a PermissionPolicy controls access of an **action** **over** resources. `**over` is very flexible: any keyword argument can be passed. But if the constructor's signature is too constraining for a developer, she can just inherit `BasePermissionPolicy` and redefine the `__init__` however she wants. 

As for `*args` I am not a fan of them for the reason we saw in `invenio_access`.
I will leave them as is for now.

ppanero commented on invenio_records_permissions/permissions/base.py:
```diff
@@ -6,99 +6,108 @@
-        self.config = config
-        self.permission_list = self.config.get_permission_list(action)
+        # Question: should this default to [Deny()] or []?
+        return [Deny()]
```
Since the configuration above returns `Deny`, and this would be an `Unknown` action, I would prefer to be safe than sorry and `Deny` it.

fenekku replied:
If we change the ones above to just `[]`, this should also be `[]`, right?

ppanero commented on invenio_records_permissions/permissions/base.py:
```diff
@@ -6,99 +6,108 @@
+        excludes = [
+            generator.excludes(**self.over) for generator in self.generators
+        ]
+        excludes = set(chain.from_iterable(excludes))
```
As soon as https://github.com/inveniosoftware/invenio-access/pull/166 gets merged, the we can use `explicit_excludes` coming from the parent class.

Something like: ``self.explicit_excludes |= excludes``

fenekku replied:
Yes. I will make another PR for that once it is merged.

ppanero commented on invenio_records_permissions/permissions/deposits.py:
```diff
@@ -6,47 +6,53 @@
 - Read access given to deposit owners.
 - Update access given to deposit owners.
 - Delete access given to admins only.
+- Use same "get_permission_policy pattern as records.py
```
Close quotes :)

fenekku replied:
oups :smile_cat:

ppanero commented on setup.py:
```diff
@@ -21,9 +21,11 @@
     'pytest-invenio>=1.0.5',
-    'invenio-accounts>=1.1.1,<1.2.0'
+    'invenio-accounts>=1.1.1,<1.2.0',
+    'Sphinx>=1.5.1',
```
I understand sphinx gets called in the tests. But is required in `docs`. This is a common issue accross all Invenio. We should start fixing it in all modules. However, Im not in favor of setting a second dependency here. Maybe we can extract it to a variable or append the `tests` array with `docs` (to avoid inconsistencies).

fenekku replied:
Agreed

ppanero commented on tests/test_permissions_base.py:
```diff
@@ -6,47 +6,48 @@
+    assert isinstance(policy(action='read').generators[0], AnyUser)
+    assert isinstance(policy(action='update').generators[0], Deny)
+    assert isinstance(policy(action='delete').generators[0], Deny)
+    assert isinstance(policy(action='random').generators[0], Deny)
```
Missing `foo_bar`?

fenekku replied:
:heavy_check_mark:

ppanero commented on tests/test_permissions_base.py:
```diff
@@ -56,14 +57,17 @@ def test_base_permission():
     # FIXME: will fail because invenio-access adds all in 'needs'
     # https://github.com/inveniosoftware/invenio-access/issues/165
-    assert update_perm.excludes == {any_user}
+    # assert update_perm.excludes == {any_user}
```
Should be okay after: https://github.com/inveniosoftware/invenio-access/pull/166 Maybe is better if we allow the tests to fail, but we uncomment, in order to notice it.

fenekku replied:
I'd rather merge only passing tests. I will remember :elephant:

ppanero commented on tests/test_permissions_base.py:
```diff
@@ -56,14 +57,17 @@ def test_base_permission():
     # FIXME: will fail because invenio-access adds all in 'needs'
     # https://github.com/inveniosoftware/invenio-access/issues/165
-    assert delete_perm.excludes == {any_user}
+    # assert delete_perm.excludes == {any_user}
```
Should be okay after: https://github.com/inveniosoftware/invenio-access/pull/166 Maybe is better if we allow the tests to fail, but we uncomment, in order to notice it.

fenekku replied:
same as above

ppanero commented on tests/test_permissions_records.py:
```diff
@@ -29,23 +32,33 @@
 
     assert create_perm.needs == {superuser_access}
-    # FIXME: will fail because invenio-access adds all in 'needs'
+    # Fails because of bug in invenio-access:
```
Should be okay after: https://github.com/inveniosoftware/invenio-access/pull/166

fenekku commented:
I have removed the commented out sections, shortened the commit message and squashed. Take another look at some of the remaining questions. We can go over things tomorrow if you would like.

fenekku commented on .travis.yml:
```diff
@@ -1,10 +1,13 @@
 #
-# Copyright (C) 2019 CERN.
+# This file is part of Invenio.
+# Copyright (C) 2015-2018 CERN.
```
Thanks! I am using `Northwestern University` only for now until Kristi gets back to me about copyright (I asked her and she needs to check with legal).

fenekku commented:
@ppanero Travis is finally passing! I reverted the multi_line_output to wait on resolution from the other PR. We can discuss the `Deny` situation on Tuesday, but I would be in favour of merging this in as-is and revisiting the issue in another PR. Thanks!

fenekku commented on .travis.yml:
```diff
@@ -1,10 +1,17 @@
-# Invenio-Records-Permissions is free software; you can redistribute it
-# and/or modify it under the terms of the MIT License; see LICENSE file for
-# more details.
+# Invenio is free software; you can redistribute it and/or modify it
```
That's my bad, I did a diff comparison with another .travis.yml file and took too much from it. Thanks!

fenekku commented on .travis.yml:
```diff
@@ -1,10 +1,17 @@
 #
+# This file is part of Invenio.
 # Copyright (C) 2019 CERN.
+# Copyright (C) 2019 Northwestern University.
```
Unfortunately, like I said I am waiting for @kristiholmes on that one. I will take care of changing copyrights once she provides me with the right one.

fenekku commented on invenio_records_permissions/factories/deposits.py:
```diff
@@ -0,0 +1,37 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2019 CERN.
+# Copyright (C) 2019 Northwestern University.
```
Same here. I will also do changes in the other module once this is settled.
